qiskit_experiments/framework/transpilation.py

```diff
@@ -0,0 +1,217 @@
+    :class:`~qiskit.transpiler.PassManager` and does not fill the circuit with
+    ancillas.
+
+    Args:
+        circuit: The :class:`~qiskit.QuantumCircuit` to re-index.
+        physical_qubits: The list of new indices for ``circuit``'s qubit indices.
+        n_qubits: Optional qubit size to use for the output circuit. If
+            ``None``, then the maximum of ``physical_qubits`` will be used.
+
+    Returns:
+        The quantum circuit with new qubit indices
+    """
+    if len(physical_qubits) != circuit.num_qubits:
+        raise QiskitError(
+            f"Circuit to map has {circuit.num_qubits} qubits, but "
+            f"{len(physical_qubits)} physical qubits specified for mapping."
+        )
+
+    # if all(p == r for p, r in zip(physical_qubits, range(circuit.num_qubits))):
+    #     # No mapping necessary
+    #     return circuit
+
+    circ_size = n_qubits if n_qubits is not None else (max(physical_qubits) + 1)
+    p_qregs = QuantumRegister(circ_size)
+    p_circ = QuantumCircuit(
+        p_qregs,
+        *circuit.cregs,
+        name=circuit.name,
+        metadata=circuit.metadata,
+        global_phase=circuit.global_phase,
+    )
+    p_circ.compose(
+        circuit,
+        qubits=physical_qubits,
+        inplace=True,
+        copy=False,
+    )
+    return p_circ
+
+
+def _has_calibration(target: Target, name: str, qubits: tuple[int, ...]) -> bool:
+    """Wrapper to work around bug in Target.has_calibration"""
+    try:
+        has_cal = target.has_calibration(name, qubits)
+    except AttributeError:
+        has_cal = False
+
+    return has_cal
+
+
+def check_transpilation_needed(
+    circuits: Sequence[QuantumCircuit],
+    backend: Backend,
+) -> bool:
+    """Test if circuits are already compatible with backend
+
+    This function checks if circuits are able to be executed on ``backend``
+    without transpilation. It loops through the circuits to check if any gate
+    instructions are not included in the backend's
+    :class:`~qiskit.transpiler.Target`.  The :class:`~qiskit.transpiler.Target`
+    is also checked for custom pulse gate calibrations for circuit's
+    instructions.  If all gates are included in the target and there are no
+    custom calibrations, the function returns ``False`` indicating that
+    transpilation is not needed.
+
+    This function returns ``True`` if the version of ``backend`` is less than
+    2.
+
+    The motivation for this function is that when no transpilation is necessary
+    it is faster to check the circuits in this way than to run
+    :func:`~qiskit.transpile` and have it do nothing.
+
+    Args:
+        circuits: The circuits to prepare for the backend.
+        backend: The backend for which the circuits should be prepared.
+
+    Returns:
+        ``True`` if transpilation is needed. Otherwise, ``False``.
+    """
+    transpilation_needed = False
+
+    if getattr(backend, "version", 0) <= 1:
+        # Fall back to transpilation for BackendV1
```

Comment:
Why this always requires full transpile?

Reply:
We could relax this condition if we want. For simplicity, I just wanted to use a `Target` below. Since the fall back case is just the current status, it is not a regression to ignore `BackendV1`. We could instead do the `BackendV1` to `BackendV2` conversion or some other approach. Partly I was thinking that that backend conversion occasionally hits edge cases and also Qiskit wants to drop `BackendV1` in 2.0, I think, so it seemed more future proof not handle `BackendV1`, but I don't feel strongly.